server/src/main/java/org/elasticsearch/action/bulk/SimulateBulkRequest.java

```diff
@@ -84,4 +125,35 @@ public Map<String, Map<String, Object>> getPipelineSubstitutions() {
     public boolean isSimulated() {
         return true;
     }
+
+    @Override
+    public Map<String, ComponentTemplate> getComponentTemplateSubstitutions() {
+        if (componentTemplateSubstitutions == null) {
+            return Map.of();
+        }
+        Map<String, ComponentTemplate> result = new HashMap<>(componentTemplateSubstitutions.size());
+        for (Map.Entry<String, Map<String, Object>> rawEntry : componentTemplateSubstitutions.entrySet()) {
+            result.put(rawEntry.getKey(), convertRawTemplateToComponentTemplate(rawEntry.getValue()));
+        }
+        return result;
+    }
+
+    private static ComponentTemplate convertRawTemplateToComponentTemplate(Map<String, Object> rawTemplate) {
+        ComponentTemplate componentTemplate;
+        try (var parser = XContentHelper.mapToXContentParser(XContentParserConfiguration.EMPTY, rawTemplate)) {
+            componentTemplate = ComponentTemplate.parse(parser);
+        } catch (IOException e) {
+            throw new AssertionError(e);
```

Comment:
I don't think we want to throw an assertion error here, as that would kill the ES node, we can probably throw an exception that will fail the simulation though?

Reply:
Oops. This was not supposed to be an AssertionError!